pkg/compiler/check.go

```diff
@@ -861,14 +869,24 @@ func (comp *compiler) checkTypeArgs(t *ast.Type, desc *typeDesc, flags checkFlag
 	return args
 }
 
-func (comp *compiler) replaceTypedef(ctx *checkCtx, t *ast.Type, flags checkFlags) {
+func (comp *compiler) replaceTypedef(ctx *checkCtx, t *ast.Type, flags *checkFlags) {
 	typedefName := t.Ident
 	comp.usedTypedefs[typedefName] = true
 	if len(t.Colon) != 0 {
 		comp.error(t.Pos, "type alias %v with ':'", t.Ident)
 		return
 	}
 	typedef := comp.typedefs[typedefName]
+	// Handling optional BASE argument.
+	if len(typedef.Args) > 0 && typedef.Args[0].Name == "BASE" {
+		*flags |= checkIsTypedefWithBase
+		if len(t.Args) == 0 {
```

Comment:
Is this check correct? I would expect something like len(t.Args) == len(typedef.Args) - 1?
For builtin types we don't add base unconditionally. Use of base depends on context, for syscall args user must not provide base, for fields they must provide base (it does not default to intptr for fields).
We may need more tests.

Reply:
Fixed the check. I've added a few more test cases.

I need to check the code on builtin types' bases.

Reply:
Okay, I've changed this to try and follow the behavior of builtin types' bases. Base argument for templates should only be provided for fields and defaults to `intptr` for syscall arguments. I've added a couple of new tests to show this too.